oximeter/collector/src/agent.rs

```diff
@@ -0,0 +1,612 @@
+                    producer, updating collection information";
+                   "producer_id" => id.to_string(),
+                   "interval" => ?info.interval,
+                   "address" => info.address,
+                );
+                value.get_mut().0 = info.clone();
+                value
+                    .get()
+                    .1
+                    .inbox
+                    .send(CollectionMessage::Update(info))
+                    .await
+                    .unwrap();
+            }
+        }
+        Ok(())
+    }
+
+    /// Forces a collection from all producers.
+    ///
+    /// Returns once all those values have been inserted into Clickhouse,
+    /// or an error occurs trying to perform the collection.
+    pub async fn force_collection(&self) {
+        let mut collection_oneshots = vec![];
+        let collection_tasks = self.collection_tasks.lock().await;
+        for (_id, (_endpoint, task)) in collection_tasks.iter() {
+            let (tx, rx) = oneshot::channel();
+            // Scrape from each producer, into oximeter...
+            task.inbox.send(CollectionMessage::Collect(tx)).await.unwrap();
+            // ... and keep track of the token that indicates once the metric
+            // has made it into Clickhouse.
+            collection_oneshots.push(rx);
+        }
+        drop(collection_tasks);
+
+        // Only return once all producers finish processing the token we
+        // provided.
+        //
+        // NOTE: This can either mean that the collection completed
+        // successfully, or an error occurred in the collection pathway.
+        futures::future::join_all(collection_oneshots).await;
+    }
+
+    /// List existing producers.
+    pub async fn list_producers(
+        &self,
+        start_id: Option<Uuid>,
+        limit: usize,
+    ) -> Vec<ProducerEndpoint> {
+        let start = if let Some(id) = start_id {
+            Bound::Excluded(id)
+        } else {
+            Bound::Unbounded
+        };
+        self.collection_tasks
+            .lock()
+            .await
+            .range((start, Bound::Unbounded))
+            .take(limit)
+            .map(|(_id, (info, _t))| info.clone())
+            .collect()
+    }
+
+    /// Delete a producer by ID, stopping its collection task.
+    pub async fn delete_producer(&self, id: Uuid) -> Result<(), Error> {
+        let (_info, task) = self
+            .collection_tasks
+            .lock()
+            .await
+            .remove(&id)
+            .ok_or_else(|| Error::NoSuchProducer(id))?;
+        debug!(
+            self.log,
+            "removed collection task from set";
+            "producer_id" => %id,
+        );
+        match task.inbox.send(CollectionMessage::Shutdown).await {
+            Ok(_) => debug!(
+                self.log,
+                "shut down collection task";
+                "producer_id" => %id,
+            ),
```

Comment:
Perhaps even making it a required field? :thinking: This way if something gets deleted, another person can always go and find out why.

Reply:
The only reason we ever delete a producer is because Nexus told us to. In fact, today _nothing_ deletes a producer (e.g., see #3808), which is its own bug.

Reply:
ahhh gotcha, thanks for the explanation!